test_util/src/lib.rs

```diff
@@ -108,44 +123,242 @@ pub fn generate_aead_key() -> LessSafeKey {
+                break;
+            }
+        }
     }
 }
 
+#[async_trait]
 impl Clock for MockClock {
+    type Interval = MockInterval;
+
     fn now(&self) -> Time {
-        let current_time = self.current_time.lock().unwrap();
-        *current_time
+        let current_times = self.inner.current_times_std_mutex.lock().unwrap();
+        current_times.0
+    }
+
+    fn now_monotonic(&self) -> Instant {
+        let current_times = self.inner.current_times_std_mutex.lock().unwrap();
+        current_times.1
+    }
+
+    async fn timeout<O, F>(&self, duration: StdDuration, future: F) -> Result<O, Elapsed>
+    where
+        F: Future<Output = O> + Send,
+    {
+        let deadline = self.now_monotonic() + duration;
+        MockTimeoutCancellable::new(future, || self.now_monotonic() >= deadline).await
+    }
+
+    fn interval_at(&self, start: Instant, period: StdDuration) -> Self::Interval {
+        MockInterval {
+            next_tick: start,
+            period,
+            clock: self.clone(),
+        }
+    }
+
+    async fn sleep(&self, duration: StdDuration) {
+        let deadline = self.now_monotonic() + duration;
+        self.sleep_until(deadline).await;
     }
 }
 
 impl Default for MockClock {
     fn default() -> Self {
-        Self {
-            // Sunday, September 9, 2001 1:46:40 AM UTC
-            current_time: Arc::new(Mutex::new(Time::from_seconds_since_epoch(1000000000))),
+        // Sunday, September 9, 2001 1:46:40 AM UTC
+        Self::new(Time::from_seconds_since_epoch(1000000000))
+    }
+}
+
+/// An interval used in conjunction with the [`MockClock`].
+#[derive(Debug)]
+pub struct MockInterval {
+    next_tick: Instant,
+    period: StdDuration,
+    clock: MockClock,
+}
+
+#[async_trait]
+impl ClockInterval for MockInterval {
+    async fn tick(&mut self) {
+        self.clock.sleep_until(self.next_tick).await;
+        self.next_tick += self.period;
+    }
+
+    fn set_missed_tick_behavior(&mut self, _behavior: MissedTickBehavior) {
+        // Ignore this setting, because time doesn't advance smoothly with a MockClock.
+    }
+}
+
+/// This future combinator is used by `MockClock::timeout` to check the timeout at every yield
+/// point, and cancel if the timeout has elapsed.
+#[pin_project]
+struct MockTimeoutCancellable<FU, FN, O>
+where
+    FU: Future<Output = O>,
+{
+    #[pin]
+    inner: FU,
+    should_cancel: FN,
+}
+
+impl<FU, FN, O> MockTimeoutCancellable<FU, FN, O>
+where
+    FU: Future<Output = O>,
+{
+    fn new(future: FU, should_cancel_predicate: FN) -> MockTimeoutCancellable<FU, FN, O> {
+        MockTimeoutCancellable {
+            inner: future,
+            should_cancel: should_cancel_predicate,
         }
     }
 }
 
+impl<FU, FN, O> Future for MockTimeoutCancellable<FU, FN, O>
+where
+    FU: Future<Output = O>,
+    FN: Fn() -> bool,
+{
+    type Output = Result<O, Elapsed>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
```

Comment:
(this comment only relevant if we keep `MockTimeoutCancellable` around)

I think this `poll` implementation will behave poorly if `inner` deadlocks: if the `inner` future never becomes ready to be polled again (i.e. never wakes the waker in `cx`), nothing will ever trigger a call to this `poll` method, so we will never notice when the timeout occurs.

I think we want to store two internal futures: the current `inner` future containing business logic, and a second future that is driven by the clock advancing and completes when the timeout occurs. I think each call to this `poll` method would poll the "timeout" future first, then the `inner` future if we haven't timed out.